_posts/2018-11-25-rafting-for-fun-and-profit.md

```diff
@@ -0,0 +1,442 @@
+    td.replace(raftStateMachine, 'sendAllHeartbeats', td.function());
+
+    // when
+    await raftStateMachine.transitionToCandidate();
+
+    // then
+    assert.equal(raftStateMachine.state.role, roles.leader);
+    assert.equal(raftStateMachine.state.currentTerm, 2);
+});
+```
+
+And the implementation:
+```javascript
+   leaderTimedOut() {
+        const now = new Date().getTime();
+        const diff = Math.abs(now - this.state.lastHeartbeat);
+        const result = diff > config.leaderTimeout
+
+        return result;
+    }
+
+    async transitionToCandidate() {
+        if (this.leaderTimedOut() && this.state.role !== roles.leader) {
+            this.state.role = roles.candidate;
+            this.state.currentTerm += 1;
+
+            const votes = await this.getVotes();
+            const positive = this.countPositiveVotes(votes);
+            logger.log('received positive votes', positive);
+            this.state.votesGranted = positive;
+
+            if (positive > (this.state.peers.size / 2)) {
+                this.transitionToLeader(); // (!)
+            } else {
+                this.resetLeaderTimeout(); // (!)
+            }
+        }
+    }
+```
+
+I marked transition changes with (!) comment.
+
+So far so good, now let's dig into something even more complex:
+
+
+```javascript
+it('should transition to candidate role when it did not receive a heartbeat', async function() {
+});
+```
+
+Hmm - this one is more tricky, we need some timeouts
+and keep track of the last heartbeat.
+I added a `setTimeout` coupled with `clearTimeout`,
+a function that will check if leader timed out - `leaderTimedOut()`
+and voila: it works...
+
+```javascript
+
+class RaftStateMachine {
+    constructor({ nodeId }) {
+        if (nodeId == undefined) {
+            throw new Error('nodeId and peers must be defined, peers must be an array');
+        }
+
+        this.state = { ...emptyState, peers: new Set(), nodeId };
+        this.alreadyVotedInThisTerm = false;
+        this.resetLeaderTimeout();
+    }
+    // ...
+
+    async transitionToCandidate() {
+        if (this.leaderTimedOut() && this.state.role !== roles.leader) {
+            this.state.role = roles.candidate;
+            this.state.currentTerm += 1;
+            this.state.alreadyVotedInThisTerm = true;
+
+            const votes = await this.getVotes();
+            const positive = this.countPositiveVotes(votes);
+            logger.log('received positive votes', positive);
+            this.state.votesGranted = positive;
+
+            if (positive > (this.state.peers.size / 2)) {
+                clearTimeout(this.leaderTimeout);
+                this.transitionToLeader();
+            } else {
+                this.resetLeaderTimeout(); // (1)
+            }
+        }
+    }
+
+    resetLeaderTimeout() {
+        clearTimeout(this.leaderTimeout);
+        this.leaderTimeout = setTimeout(this.transitionToCandidate.bind(this), config.leaderTimeout);
```

Comment:
Owinięcie w lambdę chyba też zdałoby egzamin? Ten `.bind(this)` kojarzy mi się z początkami Reacta :)